kamon-akka-http-server/src/main/scala/kamon/akka/http/server/instrumentation/ServerInstrumentation.scala

```diff
@@ -0,0 +1,158 @@
+ * limitations under the License.
+ * ========================================================== */
+package kamon.akka.http.server.instrumentation
+
+import akka.http.scaladsl.Http.{ IncomingConnection, ServerBinding }
+import akka.http.scaladsl.model.headers.RawHeader
+import akka.http.scaladsl.model.{ HttpResponse, HttpRequest }
+import akka.http.scaladsl.server.RequestContext
+import kamon.Kamon
+import kamon.akka.http.server.{ AkkaHttpServerExtension, AkkaHttpServer, AkkaHttpServerMetrics }
+import kamon.metric.Entity
+import kamon.trace.{ TraceContextAware, Tracer }
+import org.aspectj.lang.ProceedingJoinPoint
+import org.aspectj.lang.annotation._
+
+@Aspect
+class HttpRequestInstrumentation extends ServerInstrumentationUtils {
+
+  @DeclareMixin("akka.http.scaladsl.model.HttpRequest")
+  def mixinTraceContextAwareToRequest: TraceContextAware = TraceContextAware.default
+
+  @Pointcut("execution(akka.http.scaladsl.model.HttpRequest.new(..)) && this(request)")
+  def httpRequestCreation(request: HttpRequest with TraceContextAware): Unit = {}
+
+  @After("httpRequestCreation(request)")
+  def afterHttpRequestCreation(request: HttpRequest with TraceContextAware): Unit = {
+    val akkaHttpServerExtension = Kamon(AkkaHttpServer)
+    val incomingContext = Tracer.currentContext
+
+    if (incomingContext.isEmpty) {
+      val defaultTraceName = akkaHttpServerExtension.generateTraceName(request)
+      val token = request.headers.find(_.name == akkaHttpServerExtension.settings.traceTokenHeaderName).map(_.value)
+
+      val newContext = Kamon.tracer.newContext(defaultTraceName, token)
+      Tracer.setCurrentContext(newContext)
+    }
+
+    request.traceContext
+
+    recordAkkaHttpServerMetrics(request, akkaHttpServerExtension)
+  }
+}
+
+@Aspect
+class RequestContextInstrumentation extends ServerInstrumentationUtils {
+
+  @DeclareMixin("akka.http.scaladsl.server.RequestContextImpl")
+  def mixinTraceContextAwareToRequestContext: TraceContextAware = TraceContextAware.default
+
+  @Pointcut("execution(akka.http.scaladsl.server.RequestContextImpl.new(..)) && this(context)")
+  def requestContextCreation(context: RequestContext with TraceContextAware): Unit = {}
+
+  @Around("requestContextCreation(context)")
+  def afterCreation(pjp: ProceedingJoinPoint, context: RequestContext with TraceContextAware): Unit = {
+    val incomingContext = Tracer.currentContext
+
+    Tracer.withContext(incomingContext) {
+      pjp.proceed()
+    }
+  }
+}
+
+@Aspect
+class HttpResponseInstrumentation extends ServerInstrumentationUtils {
+
+  @DeclareMixin("akka.http.scaladsl.model.HttpResponse")
+  def mixinTraceContextAwareToResponse: TraceContextAware = TraceContextAware.default
+
+  @Pointcut("call(akka.http.scaladsl.model.HttpResponse.new(..)) && !cflow(execution(* *..attachTraceTokenHeader(..)))")
+  def httpResponseCreation(): Unit = {}
+
+  @Around("httpResponseCreation()")
+  def afterHttpResponseCreation(pjp: ProceedingJoinPoint) = {
```

Comment:
Here, we have a similar situation to the HttpRequest counterpart.. probably this is not the best place to do this because it might get mixed with code called when the http client side receives a response for a given request.